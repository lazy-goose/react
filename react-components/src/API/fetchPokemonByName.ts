import PokeAPI from 'pokeapi-typescript';

export async function fetchPokemonByName(name: string) {
  if (!name?.length) {
    throw new Error('Unknown name');
  }
  return await PokeAPI.Pokemon.resolve(name);
}
