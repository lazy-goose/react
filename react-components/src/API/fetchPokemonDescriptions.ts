import PokeAPI, { IPokemon } from 'pokeapi-typescript';

export async function fetchPokemonDescriptions(names: string[]) {
  const pokemonPromises = names.map((n) => PokeAPI.Pokemon.resolve(n));
  const settledPokemons = await Promise.allSettled(pokemonPromises);
  const fulfilledPokemons = settledPokemons.filter(
    (p) => p.status === 'fulfilled'
  ) as PromiseFulfilledResult<IPokemon>[];
  return fulfilledPokemons.map((p) => p.value);
}
