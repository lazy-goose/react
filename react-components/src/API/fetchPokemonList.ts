import PokeAPI, { INamedApiResource, IPokemon } from 'pokeapi-typescript';

export async function fetchPokemonList() {
  return (await PokeAPI.Pokemon.listAll()).results;
}

export type PokemonList = INamedApiResource<IPokemon>[];
