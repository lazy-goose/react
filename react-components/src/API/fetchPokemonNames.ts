import PokeAPI from 'pokeapi-typescript';

export async function fetchPokemonNames() {
  const resourceList = await PokeAPI.Pokemon.listAll();
  return resourceList.results.map((r) => r.name);
}
