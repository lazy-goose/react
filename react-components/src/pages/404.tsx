export default function Custom404() {
  return (
    <div style={{ fontSize: '4rem' }}>
      <p>Page not found</p>
    </div>
  );
}
