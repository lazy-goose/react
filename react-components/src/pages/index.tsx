import MainLayout from '@/components/MainLayout/MainLayout';
import PokeSearch from '@/components/MainLayout/PokeSearch/PokeSearch';
import {
  InferGetServerSidePropsType,
  GetServerSideProps,
  GSSPContext,
} from '@/share/types/GSSPUtils';
import { pokemonApi, store } from '@/redux';
import { INamedApiResource, IPokemon } from 'pokeapi-typescript';

export enum Query {
  Search = 'search',
  Page = 'page',
  PageSize = 'pageSize',
}

export const DEFAULT_PAGE = 1;
export const DEFAULT_PAGE_SIZE = 25;

const inMemoryCache = {
  pokemonsList: null as null | INamedApiResource<IPokemon>[],
};

export const getPokeSearchProps = async (context: GSSPContext) => {
  const pokemonName = (context.query.name as string) || '';

  const searchTerm = context.query[Query.Search] || '';
  const page = Number(context.query[Query.Page] || DEFAULT_PAGE);
  const pageSize = Number(context.query[Query.PageSize]);

  if (!inMemoryCache.pokemonsList) {
    const { data: list = [] } = await store.dispatch(
      pokemonApi.endpoints.getPokemonsList.initiate()
    );
    inMemoryCache.pokemonsList = list;
  }

  const { data: [pokemonList, total] = [[], 0] } = await store.dispatch(
    pokemonApi.endpoints.getPokemons.initiate({
      search: searchTerm,
      list: inMemoryCache.pokemonsList,
      page,
      limit: pageSize || DEFAULT_PAGE_SIZE,
    })
  );

  return {
    searchTerm,
    pageSize,
    total,
    page,
    pokemonList,
    pokemonName,
  };
};

export const getServerSideProps = (async (context) => {
  const searchProps = await getPokeSearchProps(context);
  return {
    props: {
      searchProps,
    },
  };
}) satisfies GetServerSideProps<{
  searchProps: Awaited<ReturnType<typeof getPokeSearchProps>>;
}>;

export default function RootComponent({
  searchProps,
}: InferGetServerSidePropsType<typeof getServerSideProps>) {
  return <MainLayout main={<PokeSearch {...searchProps} />} />;
}
