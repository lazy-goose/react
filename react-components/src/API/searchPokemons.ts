import PokeAPI, { IPokemon } from 'pokeapi-typescript';
import { PokemonList } from './fetchPokemonList';

const randomBetween = (min: number, max: number) =>
  Math.floor(Math.random() * (max - min + 1) + min);

export async function searchPokemons(
  value: string,
  list: PokemonList,
  page = 1,
  limit = 100
) {
  const FROM = Math.floor((page - 1) * limit);
  const TO = FROM + limit;

  const filtered = list
    .map((i) => i.name)
    .filter((n) => n.startsWith(value.toLowerCase().trimEnd()));

  const pageLimited = filtered.slice(FROM, TO);

  if (!pageLimited.length) {
    return new Promise<[[], 1]>((resolve) =>
      setTimeout(() => resolve([[], 1]), randomBetween(200, 500))
    );
  }

  const descriptionPromises = pageLimited.map((n) =>
    PokeAPI.Pokemon.resolve(n)
  );
  const descriptions = (await Promise.allSettled(descriptionPromises)).filter(
    (p) => p.status === 'fulfilled'
  ) as PromiseFulfilledResult<IPokemon>[];

  return [descriptions.map((d) => d.value), filtered.length] as const;
}
