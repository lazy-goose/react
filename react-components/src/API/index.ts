import { fetchPokemonNames } from './fetchPokemonNames';
import { searchPokemons } from './searchPokemons';
import { fetchPokemonByName } from './fetchPokemonByName';

export { fetchPokemonNames, searchPokemons, fetchPokemonByName };
