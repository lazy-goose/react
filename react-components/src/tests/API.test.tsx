import {
  vi,
  describe,
  test,
  expect,
  beforeAll,
  afterEach,
  afterAll,
  Mock,
} from 'vitest';
import { setupServer } from 'msw/node';
import { http, HttpResponse } from 'msw';
import { fetchPokemonByName, fetchPokemonList, searchPokemons } from '../API';
import pokemons from './data/pokemons.json';
import names from './data/names.json';
import pikachu from './data/pikachu.json';
import { getServerSideProps as getRootPageSSP } from '@/pages/index';
import { getServerSideProps as getPokemonPageSSP } from '@/pages/index';
import { GSSPContext } from '@/share/types/GSSPUtils';

let calls: Mock[] = [];

const POKE_URL = 'https://pokeapi.co/api/v2';

enum Calls {
  Pikachu,
  PokemonList,
  Pokemon,
}

const pushCall = (callWith: Calls) => {
  const func = vi.fn();
  func(callWith);
  calls.push(func);
};

const escapeString = (s: string) => s.replace(/[/\-\\^$*+?.()|[\]{}]/g, '\\$&');

const server = setupServer(
  http.get(`${POKE_URL}/pokemon/pikachu`, () => {
    pushCall(Calls.Pikachu);
    return HttpResponse.json(pikachu);
  }),
  http.get(`${POKE_URL}/pokemon`, () => {
    pushCall(Calls.PokemonList);
    return HttpResponse.json(pokemons);
  }),
  http.get(new RegExp(`${escapeString(POKE_URL)}/.+`), () => {
    pushCall(Calls.Pokemon);
    return HttpResponse.json(pikachu);
  })
);

describe('Test API layer', () => {
  beforeAll(() => {
    server.listen({ onUnhandledRequest: 'error' });
  });
  afterEach(() => {
    server.resetHandlers();
    calls = [];
  });
  afterAll(() => {
    server.close();
  });

  test(`fetchPokemonByName('pikachu') calls ${POKE_URL}/pokemon/pikachu`, async () => {
    await fetchPokemonByName('pikachu');
    expect(calls.length).toEqual(1);
    expect(calls.at(0)).toBeCalledWith(Calls.Pikachu);
  });

  test(`fetchPokemonList() calls ${POKE_URL}/pokemon`, async () => {
    await fetchPokemonList();
    expect(calls.length).toBeLessThanOrEqual(2);
    calls.forEach((call) => {
      expect(call).toBeCalledWith(Calls.PokemonList);
    });
  });

  test(`searchPokemons() makes request for each entry ${POKE_URL}/pokemon/{name}`, async () => {
    const pageLimit = 5;
    await searchPokemons('', names, 1, pageLimit);
    expect(calls.length).toBe(pageLimit);
    calls.forEach((call) => {
      expect(call).toBeCalledWith(Calls.Pokemon);
    });
  });

  test(`getServerSideProps() for router '/' matches snapshot props`, async () => {
    const routeData = await getRootPageSSP({
      query: { name: '' },
    } as unknown as GSSPContext);
    expect(routeData).toMatchSnapshot();
  });

  test(`getServerSideProps() for route '/pokemon/pikachu' matches snapshot props`, async () => {
    const routeData = await getPokemonPageSSP({
      query: { name: 'pikachu' },
    } as unknown as GSSPContext);
    expect(routeData).toMatchSnapshot();
  });
});
