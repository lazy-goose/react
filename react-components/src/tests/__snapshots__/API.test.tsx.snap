// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test API layer > getServerSideProps() for route '/pokemon/pikachu' matches snapshot props 1`] = `
{
  "props": {
    "searchProps": {
      "page": 1,
      "pageSize": 25,
      "pokemonList": [],
      "pokemonName": "pikachu",
      "searchTerm": "",
      "total": 0,
    },
  },
}
`;

exports[`Test API layer > getServerSideProps() for router '/' matches snapshot props 1`] = `
{
  "props": {
    "searchProps": {
      "page": 1,
      "pageSize": 25,
      "pokemonList": [],
      "pokemonName": "",
      "searchTerm": "",
      "total": 0,
    },
  },
}
`;
